### PR TITLE
8258255: Move PtrQueue active flag to SATBMarkQueue

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change to the PtrQueue hierarchy, moving the support for
"active" queues and qset to SATBMarkQueue[Set], which is the only user of
this feature.  Other classes derived from PtrQueue[Set] currently work
around or ignore this feature.  This change removes it from consideration
entirely for those other classes.

Testing:
mach5 tier1
local (linux-x64) hotspot:tier1 with -XX:+UseShenandoahGC

In the process of doing this refactoring I noticed that some of the
vmStructs support around PtrQueue[Set] didn't get moved from G1 to shared,
and the G1 parts are incomplete.  Filed JDK-8258581 to address that.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8258255](https://bugs.openjdk.java.net/browse/JDK-8258255): Move PtrQueue active flag to SATBMarkQueue

### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to 137b3b39140db80704a04f69648fd7204741a4cc
 * [Stefan Johansson](https://openjdk.java.net/census#sjohanss) (@kstefanj - **Reviewer**) ⚠️ Review applies to dccbeb46497dd58fe9e45e7f2e130a6e08aceb64

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1820/head:pull/1820`
`$ git checkout pull/1820`

### CODE REVIEW

File: src/hotspot/share/gc/shared/satbMarkQueue.hpp [hunk 1]
```diff
@@ -138,9 +150,11 @@ class SATBMarkQueueSet: public PtrQueueSet {
   // buffer; the leading entries may be excluded due to filtering.
   bool apply_closure_to_completed_buffer(SATBBufferClosure* cl);
 
+  // When active, add obj to queue by calling enqueue_known_active.
   void enqueue(SATBMarkQueue& queue, oop obj) {
     if (queue.is_active()) enqueue_known_active(queue, obj);
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/satbMarkQueue.hpp:155]: Maybe something for a different CR and probably pre-existing: `enqueue()` in some contexts is already called with `queue.is_active() == true`. I.e. this check is superfluous, particularly when called for object arrays.

**Review by tschatzl — APPROVED** [review]
Apart from that request with the comment, lgtm.

File: src/hotspot/share/gc/shared/satbMarkQueue.hpp (hunk 1 above)
Comment (kimbarrett) [reply src/hotspot/share/gc/shared/satbMarkQueue.hpp:155]: I was planning to look at that later.  I filed a reminder: https://bugs.openjdk.java.net/browse/JDK-8258607

File: src/hotspot/share/gc/shared/satbMarkQueue.hpp [hunk 2]
```diff
@@ -95,7 +103,9 @@ class SATBMarkQueueSet: public PtrQueueSet {
   // These are rarely (if ever) changed, so same cache line as count.
   size_t _process_completed_buffers_threshold;
   size_t _buffer_enqueue_threshold;
-  DEFINE_PAD_MINUS_SIZE(2, DEFAULT_CACHE_LINE_SIZE, 3 * sizeof(size_t));
+  // SATB is only active during marking.  Enqueuing is not done when inactive.
```
[the remark this answers is not in the document]
Comment (kimbarrett) [reply src/hotspot/share/gc/shared/satbMarkQueue.hpp:106]: Sure.  Changing it to
```
  // SATB is only active during marking.  Enqueuing is only done when active.
```

**Review by kstefanj — APPROVED** [review]
Nice cleaunup, look good.

**Comment by kimbarrett** [comment]
Thanks @tschatzl and @kstefanj for reviews.

**Pull request closed without merging**